Brokerages/Fxcm/FxcmBrokerage.cs

```diff
@@ -0,0 +1,364 @@
+
+namespace QuantConnect.Brokerages.Fxcm
+{
+    /// <summary>
+    /// The FXCM brokerage implementation
+    /// </summary>
+    public partial class FxcmBrokerage : Brokerage, IDataQueueHandler, IGenericMessageListener, IStatusMessageListener
+    {
+        private readonly IOrderProvider _orderProvider;
+        private readonly IHoldingsProvider _holdingsProvider;
+        private readonly string _server;
+        private readonly string _terminal;
+        private readonly string _userName;
+        private readonly string _password;
+
+        private Thread _orderEventThread;
+        private volatile bool _orderEventThreadStopped;
+        private readonly ConcurrentQueue<OrderEvent> _orderEventQueue = new ConcurrentQueue<OrderEvent>();
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="FxcmBrokerage"/> class
+        /// </summary>
+        /// <param name="orderProvider">The order provider</param>
+        /// <param name="holdingsProvider">The holdings provider</param>
+        /// <param name="server">The url of the server</param>
+        /// <param name="terminal">The terminal name</param>
+        /// <param name="userName">The user name (login id)</param>
+        /// <param name="password">The user password</param>
+        public FxcmBrokerage(IOrderProvider orderProvider, IHoldingsProvider holdingsProvider, string server, string terminal, string userName, string password)
+            : base("FXCM Brokerage")
+        {
+            _orderProvider = orderProvider;
+            _holdingsProvider = holdingsProvider;
+            _server = server;
+            _terminal = terminal;
+            _userName = userName;
+            _password = password;
+        }
+
+        #region IBrokerage implementation
+
+        /// <summary>
+        /// Returns true if we're currently connected to the broker
+        /// </summary>
+        public override bool IsConnected
+        {
+            get
+            {
+                return _gateway != null && _gateway.isConnected();
+            }
+        }
+
+        /// <summary>
+        /// Connects the client to the broker's remote servers
+        /// </summary>
+        public override void Connect()
+        {
+            if (IsConnected) return;
+
+            Log.TraceFormat("FxcmBrokerage.Connect()");
+
+            // create new thread to fire order events in queue
+            _orderEventThread = new Thread(() =>
+            {
+                while (!_orderEventThreadStopped)
+                {
+                    OrderEvent orderEvent;
+                    if (_orderEventQueue.IsEmpty || !_orderEventQueue.TryDequeue(out orderEvent))
+                    {
+                        Thread.Sleep(1);
+                        continue;
+                    }
+
+                    OnOrderEvent(orderEvent);
+                }
+            });
+            _orderEventThread.Start();
+            while (!_orderEventThread.IsAlive) {}
```

Comment:
We may want a `Thread.Sleep(1)` in here to allow other threads a change to do something while we're spinning here. Is this expected to be less than a 1 ms wait here?